Add docs for configuring kubelet credential provider plugins (#25226)
kubernetes/website: content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md

---
title: Kubelet Image Credential Provider
reviewers:
- liggitt
- cheftako
description: Configure the kubelet's image credential provider plugin
content_type: task
---

> 

Author: done

<!-- overview -->

Starting from Kubernetes v1.20, the kubelet can dynamically retrieve credentials for a container image registry
using exec plugins. The kubelet and the exec plugin communicate through stdio (stdin, stdout, and stderr) using
Kubernetes versioned APIs. These plugins allow the kubelet to request credentials for a container registry dynamically
as oppposed to storing static credentials on disk. For example, the plugin may talk to a local metadata server to retrieve
short-lived credentials for an image that is being pulled by the kubelet.

You may be interested in using this capabilitiy if any of the below are true:

* API calls to a cloud provider service are required to retreive authentication information for a registry.
* Credentials have short expiration times and requesting new credentials frequently is required.
* Storing registry credentials on disk or in image pull Secrets is not acceptable.

This guide demonstrates how to configure the kubelet's image credential provider plugin mechanism.

## {{% heading "prerequisites" %}}

* The kubelet image credential provider is introduced in v1.20 as an alpha feature. As with other alpha features,

> a feature gate `KubeletCredentialProviders` must be explicitly enabled for the feature to work

Reviewer: _Where_ does the reader enable that feature gate? Just for the kubelet?

* A working implementation of a credential provider exec plugin. You can build your own plugin or use one provided by cloud providers.

<!-- steps -->

## Installing Plugins on Nodes

A credential provider plugin is an exectuable binary that will be run by the kubelet. Ensure that the plugin binary exists on
every node in your cluster and stored in a known directory. The directory will be required later when configuring kubelet flags.

> ## Configuring Kubelet

Reviewer: ```suggestion
## Configure the kubelet
```

> ### Flags

Reviewer: I'd remove this line.

In order to use this feature, the kubelet expects two flags to be set:
* `--image-credential-provider-config` - the path to the credential provider plugin config file.
* `--image-credential-provider-bin-dir` - the path to the directory where credential provider plugin binaries are located.

### Credential Provider Configuration File

The configuration file passed into `--image-credential-provider-config` is read by the kubelet to determine which exec plugins
should be invoked for which container images. Here's an example configuration file you may end up using if you are using the ECR-based plugin:

```yaml
kind: CredentialProviderConfig
apiVersion: kubelet.config.k8s.io/v1alpha1

> providers:

Reviewer 2: I've found it helpful to annotate config and request/response documentation with the API docs, e.g. https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#request

This is a great place to explain what each field means, what it does, etc

Author: Good call, will updat

  - name: ecr
    matchImages:
    - "*.dkr.ecr.*.amazonaws.com"
    - "*.dkr.ecr.*.amazonaws.cn"
    - "*.dkr.ecr-fips.*.amazonaws.com"
    - "*.dkr.ecr.us-iso-east-1.c2s.ic.gov"
    - "*.dkr.ecr.us-isob-east-1.sc2s.sgov.gov"
    defaultCacheDuration: "12h"
    apiVersion: credentialprovider.kubelet.k8s.io/v1alpha1
```

> 

Reviewer: Should this task list an AWS EC2 instance as a prerequisite? Alternatively, is there a different credential mechanism that we can explain, perhaps something more generic?

Author: Trying to use a real example in the docs instead of dummy values. I think AWS ECR is a good reference because you can enable it on non-ec2 instance as long as there are credentials on disk.

The `providers` field is a list of enabled plugins used by the kubelet. Each entry has a few required fields:
* `name`: the name of the plugin which MUST match the name of the executable binary that exists in the directory passed into `--image-credential-provider-bin-dir`.
* `matchImages`: a list of strings used to match against images in order to determine if this provider should be invoked. More on this below.
* `defaultCacheDuration`: the default duration the kubelet will cache credentials in-memory if a cache duration was not specified by the plugin.
* `apiVersion`: the api version that the kubelet and the exec plugin will use when communicating.

> Each credential provider can also be given optional args and environment variables like so:

Reviewer: ```suggestion
If you need to set environment variables for the credential provider plugin to work as expected, you can implement that. `.providers[].env` takes an array of environment variables.

To specify environment variables you set separate `name` and `value` properties; for example:
```

```yaml
kind: CredentialProviderConfig
apiVersion: kubelet.config.k8s.io/v1alpha1
providers:
  - name: myplugin
    matchImages:
    - "*.example.com"
    defaultCacheDuration: 12h
    apiVersion: credentialprovider.kubelet.k8s.io/v1alpha1
    args:
    - get-credentials

>     env:
>     - name: EXAMPLE_ENV_VAR_NAME
>       value: EXAMPLE_ENV_VAR_VALUE

Reviewer: ```suggestion
    env:
    - name: AWS_PROFILE
      value: example_profile
```

```

Consult the plugin implementors to determine what set of arguments and environment variables are required for a given plugin.

> #### Match Images

Reviewer: ```suggestion
#### Configure image matching
```

> The `matchImages` field for each credenetial provider is used by the kubelet to determine whether a plugin should be invoked

Reviewer 2: ```suggestion
The `matchImages` field for each credential provider is used by the kubelet to determine whether a plugin should be invoked
```

for a given image that a Pod is using. Each entry in `matchImages` is an image pattern which can optionally contain a port and a path.
Globs can be used in the domain, but not in the port or the path. Globs are supported as subdomains like `*.k8s.io` or `k8s.*.io`,
and top-level domains such as `k8s.*`. Matching partial subdomains like `app*.k8s.io` is also supported. Each glob can only match
a single subdomain segment, so `*.io` does NOT match `*.k8s.io`.

> A match exists between an image and a matchImage entry when all of the below are true:

Reviewer: ```suggestion
A match exists between an image name and a `matchImage` entry when all of the below are true:
```

* Both contain the same number of domain parts and each part matches.
* The URL path of match image must be a prefix of the target image URL path.
* If the imageMatch contains a port, then the port must match in the image as well.

> Some example values of `matchImages` are:

Reviewer: ```suggestion
Some example values of `matchImages` patterns are:
```

Starting from Kubernetes v1.20, the kubelet can dynamically retrieve credentials for a container image registry
using exec plugins. The kubelet and the exec plugin communicate through stdio (stdin, stdout, and stderr) using
Kubernetes versioned APIs. These plugins allow the kubelet to request credentials for a container registry dynamically
as oppposed to storing static credentials on disk. For example, the plugin may talk to a local metadata server to retrieve
short-lived credentials for an image that is being pulled by the kubelet.

You may be interested in using this capabilitiy if any of the below are true:
1. API calls to a cloud provider service are required to retreive authentication information for a registry.
2. Credentials have short expiration times and requesting new credentials frequently is required.
3. Storing registry credentials on disk or in image pull Secrets is not acceptable.

This guide demonstrates how to configure the kubelet's image credential provider plugin mechanism.

{{< note >}}
This capabilitiy will eventually replace built-in logic in the kubelet to serve registry credentials for Azure, AWS and GCP.
{{< /note >}}

## {{% heading "prerequisites" %}}

> * kubelet on v1.20 or later. If on v1.20, ensure the alpha feature gate `KubeletCredentialProviders` is on.

Reviewer 3: `KubeletCredentialProviders` needs to be added to the feature gate list `content/en/docs/reference/command-line-tools-reference/feature-gates/` with a short description describing the feature

Author: ack, thanks

Author: Great call @tengqm, will update with your suggestion

* A working implementation of a credential provider exec plugin. You can build your own plugin or use one provided by cloud providers.

<!-- steps -->

## Installing Plugins on Nodes

A credential provider plugin is an exectuable binary that will be run by the kubelet. Ensure that the plugin binary exists on
every node in your cluster and stored in a known directory. The directory will be required later when configuring kubelet flags.

## Configuring Kubelet

### Flags

In order to use this feature, the kubelet expects two flags to be set:
* `--image-credential-provider-config` - the path to the credential provider plugin config file.

> * `--image-credential-provider-bin-dir` - the path to the directory where credential provider plugin binaries are located.

Author: We did not add config fields for these, which I think has been expected for container runtime options. For example, none of the flags in `ContainerRuntimeOptions` have config field equivalents:

https://github.com/kubernetes/kubernetes/blob/51441fd052756d2c76daee83d39b352bd8a2d265/pkg/kubelet/config/flags.go#L26-L82

If you are on v1.20, ensure the feature gate `KubeletCredentialProviders` is enabled via the `--feature-gates` flag.

### Credential Provider Configuration File

The configuration file passed into `--image-credential-provider-config` is read by the kubelet to determine which exec plugins
should be invoked for which container images. Here's an example configuration file you may end up using if you are using the ECR-based plugin:

```yaml

> kind: CredentialProviderConfig

Author: It is a separate API group, distinct from kubelet config https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider

> kind: CredentialProviderConfig

Reviewer 4: @andrewsykim The API group you are talking about only contains `CredentialProviderRequest`, `CredentialProviderRequest` and `AuthConfig`. It does not contain `CredentialProviderConfig`. I was hoping that `CredentialProviderConfig` in the `k8s.io/kubelet/config/v1beta1/types.go`.

> kind: CredentialProviderConfig

Author: Oh my bad, CredentialProviderConfig is in the v1alpha1 kubelet config API group for v1.20.

> kind: CredentialProviderConfig

Reviewer 4: The weird thing is that `CredentialProviderConfig` only exists in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/types.go#L468, not in the staging repo: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/config/v1beta1/types.go

Is this a bug to fix?

I noticed this because I was trying to generate kubelet configuration documentation, along with other "unpublished" API types. See: https://github.com/kubernetes/website/pull/25094 and the preview is here: https://deploy-preview-25094--kubernetes-io-master-staging.netlify.app/docs/reference/command-line-tools-reference/kubelet-config.v1beta1/
`CredentialProviderConfig` is not shown on that page.

> kind: CredentialProviderConfig

Author: `CredentialProviderConfig` is in staging, but it is only in the v1alpha1 API group https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/config/v1alpha1/types.go. It will be promote to the v1beta1 group in a future release.

> kind: CredentialProviderConfig

Reviewer 4: Ah I see. Sorry for all the stupid questions. I was thinking that "if everything is up to date, it should appear in the latest version", however, that is not the case. Thanks for the clarification.

apiVersion: kubelet.config.k8s.io/v1alpha1
providers:
  - name: ecr
    matchImages:
    - "*.dkr.ecr.*.amazonaws.com"
    - "*.dkr.ecr.*.amazonaws.cn"
    - "*.dkr.ecr-fips.*.amazonaws.com"
    - "*.dkr.ecr.us-iso-east-1.c2s.ic.gov"
    - "*.dkr.ecr.us-isob-east-1.sc2s.sgov.gov"

>     defaultCacheDuration: 12h

Reviewer 4: ```suggestion
    defaultCacheDuration: "12h"
```

* `defaultCacheDuration`: the default duration the kubelet will cache credentials in-memory if a cache duration was not specified by the plugin.
* `apiVersion`: the api version that the kubelet and the exec plugin will use when communicating.

Each credential provider can also be given optional args and environment variables like so:

```yaml
kind: CredentialProviderConfig
apiVersion: kubelet.config.k8s.io/v1alpha1
providers:
  - name: myplugin
    matchImages:
    - "*.example.com"
    defaultCacheDuration: 12h
    apiVersion: credentialprovider.kubelet.k8s.io/v1alpha1
    args:
    - get-credentials
    env:
    - name: EXAMPLE_ENV_VAR_NAME
      value: EXAMPLE_ENV_VAR_VALUE
```

Consult the plugin implementors to determine what set of arguments and environment variables are required for a given plugin.

#### Match Images

The `matchImages` field for each credenetial provider is used by the kubelet to determine whether a plugin should be invoked for a given image that a Pod is using.
Each entry in `matchImages` is a pattern which can optionally contain a port and a path. Globs can be used in the domain, but not in the port or the path. Globs are

> supported as subdomains like '\*.k8s.io' or 'k8s.\*.io', and top-level domains such as 'k8s.\*'. Matching partial subdomains like 'app\*.k8s.io' is also supported. Each

Reviewer 4: ```suggestion
supported as subdomains like `*.k8s.io` or `k8s.*.io`, and top-level domains such as
`k8s.*`. Matching partial subdomains like `app*.k8s.io` is also supported. Each
```

The point here is that by wrapping these literals with backtiqs, readers can easily identify what are the patterns with NO confusion. Having backslashes in the text can thus considered harmful.

Author: Good call, will update

> glob can only match a single subdomain segment, so '\*.io' does NOT match '\*.k8s.io'.

Reviewer 4: Please consider wrap the long lines properly so that future changes can be easily tracked.

A match exists between an image and a matchImage entry when all of the below are true:
* Both contain the same number of domain parts and each part matches.

> * The URL path of an imageMatch must be a prefix of the target image URL path.

Reviewer 4: What is an "imageMatch"?

* If the imageMatch contains a port, then the port must match in the image as well.

Some example values of `matchImages` are:
* 123456789.dkr.ecr.us-east-1.amazonaws.com
* \*.azurecr.io
* gcr.io
* \*.\*.registry.io

> * foo.registry.io:8080/path

Reviewer 4: plse consider using backtiqs for these string literals.
